argoproj/argo-workflows
# fix: ArtifactGC Fails for Stopped Workflows. Fixes #11879

Fixes #11879

### Motivation



Currently, stopped workflows will happily save artifacts to a storage bucket but the artifacts won't get garbage collected after workflow deletion. The changes make it so that a stopped workflow's artifacts are properly garbage collected upon workflow deletion.

### Modifications



The `operator.go`'s call to `woc.taskResultReconciliation()` previously occurred after Artifact Garbage Collection (`ArtifactGC`) and an early return due to `if woc.wf.Labels[common.LabelKeyCompleted] == "true"`, leading to a scenerio in which garbage collection would fail upon workflow deletion. This in turn would cause workflow deletion to get blocked by the ArtifactGC finalizer. `woc.taskResultReconciliation()` is responsible for updating the node status in the workflow based on the artifact outputs saved during workflow execution with `reportOutputs()` in `executor.go`. 



### Verification



I tested the changes by adding an `e2e` test called `TestStoppedWorkflow` which uses a custom workflow with the following logic.

1) Ensure no artifact exists in the bucket.
2) Start the workflow.
3) Workflow runs and creates a test artifact.
4) Verifiy the artifact exists in the bucket.
4) The workflow submits a CLI-based stop event for itself.
5) The workflow is deleted by the test.
6) Artifacts are verified to have been deleted.
7) Finalizer is removed incase the test failed.

---

I made some additional minor tweaks to other files, most notably, `timeout = v - 30*time.Second + defaultTimeout` in `when.go` which would cause custom timeout settings to be incorrect when using the `WaitForWorkflow()` e2e testing utility.
